### PR TITLE
[BART] torch 1.0 compatibility

### PR DESCRIPTION
**Pull request by sshleifer**
This PR contains two minor fixes and one piece of cleanup for the BartModel.

**1.** Previously, Bart's encoder padding mask used -10000. to represent tokens that should be ignored, then called `masked_fill(mask.to(torch.bool), -inf)` to use the mask.

There are two problems with this:
- it's confusing to set a value to a large negative and then call `bool`. Why not just invert the mask and call `bool` immediately.
- `torch.bool` is released in pytorch 1.2, so this code breaks on earlier versions.
- Fix: let `torch.eq` make the mask the correct dtype at the beginning.

**2.** explicit use of `F.gelu` is not allowed/or broken in earlier torch versions. Let `ACT2FN` handle this logic. 

**3.** An unreachable code branch is deleted.

Supplementary Material: torch v 1.2.0 [release notes](https://github.com/pytorch/pytorch/releases/tag/v1.2.0)

### CODE REVIEW

File: src/transformers/modeling_bart.py [hunk 1]
```diff
@@ -700,22 +698,17 @@ def _cat_prev_key_padding_mask(
...
     ) -> Optional[Tensor]:
         # saved key padding masks have shape (bsz, seq_len)
-        if prev_key_padding_mask is not None and static_kv:
-            new_key_padding_mask = prev_key_padding_mask
-        elif prev_key_padding_mask is not None and key_padding_mask is not None:
-            new_key_padding_mask = torch.cat([prev_key_padding_mask.float(), key_padding_mask.float()], dim=1)
-        # During incremental decoding, as the padding token enters and
-        # leaves the frame, there will be a time when prev or current is None
-        elif prev_key_padding_mask is not None:
-            filler = torch.zeros(batch_size, src_len - prev_key_padding_mask.size(1))
```
Comment (sshleifer) [inline src/transformers/modeling_bart.py]: this branch is never reached

File: src/transformers/modeling_bart.py [hunk 2]
```diff
@@ -287,9 +290,7 @@ def forward(
         # check attention mask and invert
         if attention_mask is not None:
             assert attention_mask.dim() == 2
-
-            attention_mask = (1.0 - attention_mask.long()) * -10000.0
-            assert attention_mask.max() <= 0
+            attention_mask = attention_mask.eq(0).to(MASK_DTYPE)
```
Comment (julien-c) [inline src/transformers/modeling_bart.py]: isn't the output of `.eq()` already in the type of MASK_DTYPE? (bool or uint8 depending on the PyTorch version)

Comment (sshleifer) [reply src/transformers/modeling_bart.py]: Yes, great catch.

File: src/transformers/modeling_bart.py [hunk 3]
```diff
@@ -107,6 +107,9 @@ def _prepare_bart_decoder_inputs(
     return decoder_input_ids, decoder_attn_mask
 
 
+MASK_DTYPE = torch.uint8 if torch.__version__ < "1.2.0" else torch.bool
```
Comment (thomwolf) [inline src/transformers/modeling_bart.py]: Nice indeed! We may use this solution in other locations in the library.

**Review by thomwolf — APPROVED** [review]

**Comment by thomwolf** [comment]
One note: you can add an `activation_function` attribute in `BartConfig` defaulting to "gelu" to be used when calling the `ACT2FN`. This lets people switch to "gelu_new" if they want a different trade-off accuracy versus speed/memory.
